ssz/src/main/java/tech/pegasys/teku/ssz/backing/tree/SszNodeTemplate.java

```diff
@@ -0,0 +1,199 @@
+      return new Location(getOffset() + addOffset, getLength(), isLeaf());
+    }
+
+    public int getOffset() {
+      return offset;
+    }
+
+    public int getLength() {
+      return length;
+    }
+
+    public boolean isLeaf() {
+      return leaf;
+    }
+  }
+
+  public static SszNodeTemplate createFromType(ViewType viewType) {
+    checkArgument(viewType.isFixedSize(), "Only fixed size types supported");
+
+    return createFromTree(viewType.getDefaultTree());
+  }
+
+  // This should be CANONICAL binary tree
+  private static SszNodeTemplate createFromTree(TreeNode defaultTree) {
+    Map<Long, Location> gIdxToLoc =
+        binaryTraverse(
+            GIndexUtil.SELF_G_INDEX,
+            defaultTree,
+            new BinaryVisitor<>() {
+              @Override
+              public Map<Long, Location> visitLeaf(long gIndex, LeafNode node) {
+                Map<Long, Location> ret = new HashMap<>();
+                ret.put(gIndex, new Location(0, node.getData().size(), true));
+                return ret;
+              }
+
+              @Override
+              public Map<Long, Location> visitBranch(
+                  long gIndex,
+                  TreeNode node,
+                  Map<Long, Location> leftVisitResult,
+                  Map<Long, Location> rightVisitResult) {
+                Location leftChildLoc = leftVisitResult.get(gIdxLeftGIndex(gIndex));
+                Location rightChildLoc = rightVisitResult.get(gIdxRightGIndex(gIndex));
+                rightVisitResult.replaceAll(
+                    (idx, loc) -> loc.withAddedOffset(leftChildLoc.getLength()));
+                leftVisitResult.putAll(rightVisitResult);
+                leftVisitResult.put(
+                    gIndex,
+                    new Location(0, leftChildLoc.getLength() + rightChildLoc.getLength(), false));
+                return leftVisitResult;
+              }
+            });
+    return new SszNodeTemplate(gIdxToLoc, defaultTree);
+  }
+
+  private static List<Bytes> nodeSsz(TreeNode node) {
+    List<Bytes> sszBytes = new ArrayList<>();
+    TreeUtil.iterateLeavesData(node, LEFTMOST_G_INDEX, RIGHTMOST_G_INDEX, sszBytes::add);
+    return sszBytes;
+  }
+
+  private final Map<Long, Location> gIdxToLoc;
+  private final TreeNode defaultTree;
+  private final Map<Long, SszNodeTemplate> subTemplatesCache = new ConcurrentHashMap<>();
+
+  public SszNodeTemplate(Map<Long, Location> gIdxToLoc, TreeNode defaultTree) {
+    this.gIdxToLoc = gIdxToLoc;
+    this.defaultTree = defaultTree;
+  }
+
+  public Location getNodeSszLocation(long generalizedIndex) {
+    return gIdxToLoc.get(generalizedIndex);
+  }
+
+  public int getSszLength() {
+    return gIdxToLoc.get(SELF_G_INDEX).getLength();
+  }
+
+  public SszNodeTemplate getSubTemplate(long generalizedIndex) {
+    return subTemplatesCache.computeIfAbsent(generalizedIndex, this::calcSubTemplate);
+  }
+
+  private SszNodeTemplate calcSubTemplate(long generalizedIndex) {
```

Comment:
Would it help to have a check argument to make sure the given generalized index is in the subtree of this template?